Feat ai tools

Jermolene opened the pull request:
This plugin adds integrated LLM conversations to the TiddlyWiki platform.

It currently supports two different server backends:

* Locally running Llamafile server - LLlamafile is an open source project that lets you distribute and run LLMs as a single file. The files are large, typically 4+ gigabytes but offer reasonable performance on modern hardware, and total privacy
* OpenAI Service - OpenAI is a commercial service that offers paid APIs for accessing some of the most sophisticated LLMs that are available. OpenAI requires tokens to be purchased for API usage (this is entirely separate from ChatGPT subscriptions)

See the prerelease build at: https://tiddlywiki5-git-feat-ai-tools-jermolenes-projects.vercel.app

![image](https://github.com/Jermolene/TiddlyWiki5/assets/174761/d555a52b-951a-4c34-8d69-4845baad7e7d)

joocyNut commented:
![NoapiKey](https://github.com/user-attachments/assets/d8225c9f-7cdf-4a57-a46b-147f3917d0b5)
![error401](https://github.com/user-attachments/assets/051f2611-88e4-4819-afe6-529662ca8b5f)
I was testing the plugin and ran into a major issue. I was looking through the server tiddlers and can't find where the API key is referenced. My concerns were confirmed because I received an HTTP error 401 when sending an OpenAI request, and when importing the plugin into an empty wiki, it proved that even if an API key is given it will not provide authorization.

pmario commented:
![image](https://github.com/user-attachments/assets/c26d7747-7e7f-4840-8bfa-22477ef3f53d)

joocyNut commented:
> ![image](https://private-user-images.githubusercontent.com/374655/348362806-c26d7747-7e7f-4840-8bfa-22477ef3f53d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA4MDkyNDcsIm5iZiI6MTcyMDgwODk0NywicGF0aCI6Ii8zNzQ2NTUvMzQ4MzYyODA2LWMyNmQ3NzQ3LTdlN2YtNDg0MC04YmZhLTIyNDc3ZWYzZjUzZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcxMlQxODI5MDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02YjAyZTAzZTA2ZjgzYWRlMDUwMDhiMGI2ZDVkNjY4MzI3MWI2MGMxOGExMTZlZDg3MzE3YzJmZjljNzI4NmQzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.Lp5vDTlHXxfekX6DvtsO68ccNgWhAHFDFMZYaj2qqwQ)

I did provide an API key in the settings of the plugin. Does it work for you?

pmario commented:
Sorry - I do not have an OpenAI Key

Jermolene commented:
> when importing the plugin into an empty wiki, it proved that even if an API key is given it will not provide authorization.

Hi @joocyNut this PR includes a small core modification that is required by the "ai-tools" plugin in order to function. That means that for the moment you can only use the plugin with the special prerelease version of the core included in the [preview](https://tiddlywiki5-git-feat-ai-tools-jermolenes-projects.vercel.app/)

I will update the docs to clarify this point.

pmario commented:
I did have a look and was a bit confused.

- The system dialogue says: edit - copy - delete
- If I click **edit** I get: view - copy - delete

I did not know how to close the dialogue. I personally would prefer **done** - copy - delete

------

I am not happy with the round "send" button. None of our default buttons at the moment are round.
I would prefer border radius 3px or 4px as those in the system-dialogue

![image](https://github.com/user-attachments/assets/1dc17c74-e254-4512-8bea-c996bd931486)

pmario commented:
the "Choose an image" dropdown is almost invisible with "dark" palettes

![image](https://github.com/user-attachments/assets/834f970d-2209-4938-b7aa-15b8ba2ce811)

joocyNut commented:
Is it possible to get an empty version of the PR?

Jermolene commented:
> I did have a look and was a bit confused.
> 
> * The system dialogue says: edit - copy - delete
> * If I click **edit** I get: view - copy - delete
> 
> I did not know how to close the dialogue. I personally would prefer **done** - copy - delete

Hi @pmario I understand your confusion. This isn't the final user interface. For the moment I prefer "view" because this is not a true editor in that it does not support drafts, and thus doesn't allow users to cancel edits.

> I am not happy with the round "send" button. None of our default buttons at the moment are round. I would prefer border radius 3px or 4px as those in the system-dialogue

While consistency is usually our goal, here the goal is to make the button look distinctive. When using OpenAI, clicking that button is a billable event, and so it seems reasonable to signal that with an unusual visual presentation.

Jermolene commented:
> The "Choose an image" dropdown is almost invisible with "dark" palettes

HI @pmario I've pushed a fix in a1782b1e4eb43b4c83b95c46adedbe8bee38a262. Eventually we'll need to use palette colours.

Jermolene commented:
> Is it possible to get an empty version of the PR?

Hi @joocyNut at the moment the best approach is to save the empty custom core from https://tiddlywiki5-git-feat-ai-tools-jermolenes-projects.vercel.app/empty.html and then drag and drop the ai-tools plugin from https://tiddlywiki5-git-feat-ai-tools-jermolenes-projects.vercel.app/index.html

pmario commented on plugins/tiddlywiki/ai-tools/modules/conversations-archive-importer.js:
```diff
@@ -27,7 +27,8 @@ ConversationsArchiveImporter.prototype.import = function(widget,conversationsTit
 			tags: $tw.utils.stringifyList(["$:/tags/AI/Conversation"]),
 			created: conversationCreated,
-			modified: conversationModified
+			modified: conversationModified,
```
@Jermolene ... it seems modified an created fields are wrong

Jermolene replied:
Hi @pmario what do you mean? That they are the wrong way round? Or something else?

pmario replied:
UUPS, Sorry. GitHub showd 5 lines of code and I thought it is a header. So I did expect `modified: 20240719xxx` instead of a variable. Sorry forget about it. 

![image](https://github.com/user-attachments/assets/e184efbc-43e6-4a36-8fbc-7eda8c47a557)

linonetwo commented:
Is there some wikitext API for this plugin? I'm writing a workflow plugin like Defy or Coze, which is a visual version of langchain https://github.com/tiddly-gittly/workflow 

I care about how to redirect input from your plugin to workflow, and how to stream workflow output to your UI?

Jermolene commented:
> Is there some wikitext API for this plugin?

The API is not yet documented, but there is a generic function for retrieving the next message in an LLM conversation:

https://github.com/TiddlyWiki/TiddlyWiki5/pull/8365/files#diff-d9cb6ac4fd8984f05e3db03e83f998cbcf1e588010a2c73c4e40a2b5cc9df90a

Much of the logic is specific to the LLM, and so lives in the LLM server tiddlers:

https://github.com/TiddlyWiki/TiddlyWiki5/pull/8365/files#diff-87f1e1e8cbb23ac187209b09af4deeb5453bff27f6fe5c18cb2638471cf4988a

> I'm writing a workflow plugin like Defy or Coze, which is a visual version of langchain https://github.com/tiddly-gittly/workflow (WIP)
> 
> I care about how to redirect input from your plugin to workflow, and how to stream workflow output to your UI? Some intermediate node on workflow may search the wiki and append it to user prompt.

The plugin doesn't yet support the streaming variant of LLM APIs, but that is something I hope to work on soon. I think the approach would just be that a message would have a new "message-status" field that could hold values for "streaming" and "complete", and in the background streaming messages would periodically be updated with new text appended.

But I think you're asking about piping input and output between different components. I think that would be accomplished by passing tiddler titles around, rather in the way that the `get-llm-completion` procedure takes in its parameters the title of the tiddler that should receive the output. So the LLM would stream its response to a tiddler, and the workflow component would pick up that output to act on it.

linonetwo commented:
Thank you, so I just need to `addTiddler` to `<<resultTitlePrefix>>` at the end of every workflow, then any workflow can work with the UI in this plugin.

I have other questions:

1. tag `$:/tags/AI/CompletionServer` may suggest it support all kinds of AI, for example comfyui image AI API, will it be supported? Developer just need to put the generated image to `<<resultTitlePrefix>>` tiddler.
2. Can you move `procedure action-get-response` to another tiddler, and add a dropdown to pick different procedure to use? So I can provide several custom procedure for different workflows (Agents, as those people advertised), and user can choose which to run.
3. Since we have TiddlyWiki organization, why not put new plugins to separate repo, like in https://github.com/tiddly-gittly/ , so the release cycle of this plugin won't need to be the same as tw core? I think there will be frequent change to UI in the future.

This plugin may be a foundation of many 3rd party AI plugins, I'm happy for not needing to spend time writing UI for chatbot, and can focus on workflow part.

Jermolene commented:
> Thank you, I was asking about piping, so I just need to `addTiddler` to `<<resultTitlePrefix>>` at the end of every workflow, then any workflow can work with the UI in this plugin.

Yes, I think so. I've added some docs for the conversation format here: 58f96e779a34fb72b995ddbb335c3792a491366e

> 1. tag `$:/tags/AI/CompletionServer` may suggest it support all kinds of AI, for example comfyui image AI API, will it be supported? Developer just need to put the generated image to `<<resultTitlePrefix>>` tiddler.

Yes. It currently supports image input to Llamafile and OpenAI, and the schema allows for image responses, but that is not yet implemented.

> 2. Can you move `procedure action-get-response` to another tiddler, and add a dropdown to pick different procedure to use? So I can provide several custom procedure for different workflows (Agents, as those people advertised), and user can choose which to run.

I've refactored all the procedures and functions to be global here: ea595dfe2f5a2f8ac6290eae7952a9b88a8943de

> 3. Since we have TiddlyWiki organization, why not put new plugins to separate repo, like in https://github.com/tiddly-gittly/ , so the release cycle of this plugin won't need to be the same as tw core? I think there will be frequent change to UI in the future.

The purpose of introducing the stability badges is so that we can move to allowing the experimental core plugins to be updated on their own schedule, without waiting for a release (as long as there are no associated changes to the core).

> 4. UI to be added are: cancel request, see agent background thinking steps, cute avatar for waifu agents... I might PR to add it, or use some kinds of slots to add it from other plugin.

I expect a combination of both approaches will be needed, PRs are welcome in any case.

Another key improvement to add is streaming responses from the server, albeit I think that that would require some core modifications to be able to access SSE data in wikitext.

> This plugin may be a foundation of many 3rd party AI plugins, I'm happy for not needing to spend time writing UI and web API requests for chatbot, and can focus on workflow part.

Great, that definitely fits with the goal of pooling the resources of those interested in exploring LLMs in TiddlyWiki.

yedhukrishnagirish commented:
<img width="884" alt="Screenshot 2024-08-10 at 11 27 27" src="https://github.com/user-attachments/assets/fa4066ba-9d53-40fd-a50f-cea9860309ba">
 I pasted the API key, and it didn't run for me. Does it only work with gpt4.0?

linonetwo commented:
Some premitives inspired by https://wordware.ai and https://www.colelawrence.com/storyai :

1. `<$read-context name="xxx">  <<xxx>>  </$read-context>`
  1. when `invokeAction`, collect all text above, or up to a [section mark](https://github.com/Jermolene/TiddlyWiki5/pull/7744) 
  2. save to a variable `xxx`, simillar to `wikify` widget.
  3. Then the variable can be used by AI generation procedures that is called nested inside it.
  4. So user can create system prompt in normal wikitext in same tiddler
1. `<$suspend >` or `<$await >` that block the next `invokeAction`, until sub-widget tree or procedure inside it runs the `<$continue />` or `<$resolve />` widget. Or wait until a filter has output.
  2. when "async await" or even "promise resolve" comes out, JS get is reborn, because no more callback hell. Currently, widgets must wrap each other to perform async callback, like a callback hell
  3. when this is provided, user can write control flow of LLM agent easier
1. `<$references>` to collect missing or empty variables referneced in a wikitext in current context
  1. so before executing an agent, it can show a modal, asking user to input these variables.
  2. or feed the variable list to the prompt, let LLM ask user input natural language and extract variables, store to temp tiddlers, and `<$set>` to variables

linonetwo commented:
Would you consider add "Artifact Claude" or "Canvas ChatGPT" in this natively? WikiText DSL have huge advantage on creating custom data manipulating UI on the fly.
No need to worry about model, I will train a custom WikiText model when we have this plugin adapted.

Jermolene commented:
> Would you consider add "Artifact Claude" or "Canvas ChatGPT" in this natively? WikiText DSL have huge advantage on creating custom data manipulating UI on the fly. No need to worry about model, I will train a custom WikiText model when we have this plugin adapted.

Yes, I'd very much like to have more LLMs work out of the box. Are you able to work on them? With Gemini and a few other LLMs I've found it very hard to find the actual HTTP API documentation, they usually want people to use a Python library to interact with things.

Jermolene commented:
> Some premitives inspired by https://wordware.ai/ and https://www.colelawrence.com/storyai :

Thanks @linonetwo sorry I missed this. I think there are a few different ways we might achieve those features. I welcome your explorations of making the wikitext action language more powerful.

linonetwo commented:
Another idea, providing `$:/Actions` as tools for AI to use. Inspired by https://github.com/kookma/TW-Shiraz/issues/22

BTW I have too many open PR to manage now, I really want then to be close or merged before I create new ones, while this will require @Jermolene to share some time with me, but good side is I will be confident to write new PRs here, thanks.

aerugo commented:
Hi @Jermolene. Thanks a lot for TiddlyWiki. I'm a recension convert after I started working somewhere where I'm very limited in terms of what I can install or run in cloud services.

I am also working on a lot of AI related projects, so naturally this caught my interest.

I recently developed an AI chat bot client in JavaScript that can run from the file system, and can even be be hosted on a SharePoint for internal use in the company. While doing so I build integrations for Open AI, Anthropic, Microsoft Azure AI Foundry and Ollama. Feel free to check out my implementation here:

https://github.com/aerugo/stina/tree/main/src/api

Just thought I'd share it in case it might be helpful. I haven't gotten into developing for TiddlyWiki yet, but who knows.

Jermolene commented:
Hi @aerugo 

> Hi @Jermolene. Thanks a lot for TiddlyWiki. I'm a recent convert after I started working somewhere where I'm very limited in terms of what I can install or run in cloud services.

Thank you! TiddlyWiki really shines under repressive regimes!

> Just thought I'd share it in case it might be helpful. I haven't gotten into developing for TiddlyWiki yet, but who knows.

Thank you, it looks like you're working on similar lines but in JS. Our goal by making these primitives available in wikitext is to lower the barriers to building custom apps incorporating AI.

Jermolene commented:
Thanks for the feedback. I had far too many in-progress PRs for v5.4.0, and some of them unfortunately fell by the wayside.

As I mentioned at https://github.com/TiddlyWiki/TiddlyWiki5/pull/8966#issuecomment-3867063140 I would still very much like to improve our AI support.

I’ve added this PR to the plan for v5.5.0. A good strategy might would be to merge something relatively simple early on, and keep iteratively improving it. Being marked as experimental gives us a certain amount of freedom.

Jermolene commented:
> BTW, did you see that great AI-leveraged site to examine GitHub code https://deepwiki.com/search/what-is-initial-root-step-when_7458bcdb-a8c3-4416-ae4e-371540e34644?mode=fast Looks like a good helper for newbies devs in TW5

I have seen this, and parts of it are very impressive. It's difficult to know whether it is 100% accurate and so I'm not comfortable recommending it right now.

linonetwo commented:
@sukima There is another PR added http request message for send message action widget.

And read #8966 , it has built-in OpenAI API support, and you can add any other custom API servers.

For MCP see #9692 

And usually don't need MCP. I use VSCode to directly read the wiki to get the tiddlers, and use askQuestion tool to discuss with me, then write back the new content.
Thanks to tiddlywiki's local-first .tid files.

<img width="2444" height="1612" alt="截屏2026-02-28 16 10 10" src="https://github.com/user-attachments/assets/f94f7820-8f15-4aad-9c75-5abd87de64b4" />
